src/main/java/org/sonarlint/intellij/analysis/SonarLintJobManager.java

```diff
@@ -0,0 +1,199 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonarlint.intellij.analysis;
+
+import com.intellij.openapi.application.Application;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.components.AbstractProjectComponent;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.module.Module;
+import com.intellij.openapi.progress.ProgressManager;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.util.messages.MessageBus;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.sonarlint.intellij.issue.IssueProcessor;
+import org.sonarlint.intellij.messages.TaskListener;
+import org.sonarlint.intellij.trigger.TriggerType;
+import org.sonarlint.intellij.ui.SonarLintConsole;
+import org.sonarlint.intellij.util.SonarLintUtils;
+
+public class SonarLintJobManager extends AbstractProjectComponent {
+  private static final Logger LOGGER = Logger.getInstance(SonarLintJobManager.class);
+  private final IssueProcessor processor;
+  private final MessageBus messageBus;
+  private final JobQueue queue;
+  // used to synchronize the handling of queue and running status together
+  private final Object lock;
+  private final SonarLintStatus status;
+  private final SonarLintConsole console;
+
+  public SonarLintJobManager(Project project, IssueProcessor processor) {
+    super(project);
+    this.processor = processor;
+    this.messageBus = project.getMessageBus();
+    this.queue = new JobQueue(project);
+    this.lock = new Object();
+    this.status = SonarLintStatus.get(this.myProject);
+    this.console = SonarLintConsole.get(myProject);
+
+    messageBus.connect(project).subscribe(TaskListener.SONARLINT_TASK_TOPIC, new TaskListener() {
+      @Override public void started(SonarLintJob job) {
+        //nothing to do
+      }
+
+      @Override public void ended(SonarLintJob job) {
+        taskFinished();
+      }
+    });
+  }
+
+  public void submitAsync(Module m, Collection<VirtualFile> files, TriggerType trigger) {
+    if (console.debugEnabled()) {
+      SonarLintConsole.get(myProject).debug(String.format("[%s] %d file(s) submitted", trigger.getName(), files.size()));
+    }
+
+    SonarLintJob newJob = new SonarLintJob(m, files);
+    SonarLintJob nextJob;
+
+    synchronized (lock) {
+      try {
+        queue.queue(newJob);
+      } catch (JobQueue.NoCapacityException e) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1166)